tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ProactiveRefreshTests.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.Platforms.Features.OpenTelemetry;
+using Microsoft.Identity.Test.Common;
+using Microsoft.Identity.Test.Integration.Infrastructure;
+using Microsoft.Identity.Test.Integration.NetFx.Infrastructure;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using OpenTelemetry;
+using OpenTelemetry.Metrics;
+
+namespace Microsoft.Identity.Test.Integration.HeadlessTests
+{
+    [TestClass]
+    public class ProactiveRefreshTests
+    {
+        [TestMethod]
+        public async Task TestProactiveTokenWithTelemetry()
+        {
+            Trace.WriteLine("Add exporter to test the metrics for proactive token refresh");
+            List<Metric> exportedMetrics = new();
+            var meterProvider = Sdk.CreateMeterProviderBuilder()
+                .AddMeter(OtelInstrumentation.MeterName)
+                .AddInMemoryExporter(exportedMetrics)
+                .Build();
+
+            IConfidentialAppSettings settings = ConfidentialAppSettings.GetSettings(Cloud.Public);
+            settings.UseAppIdUri = true;
+
+            AuthenticationResult authResult;
+
+            Trace.WriteLine("Create a confidential client application with certificate.");
+            ConfidentialClientApplication confidentialApp = ConfidentialClientApplicationBuilder
+                .Create(settings.ClientId)
+                .WithAuthority(settings.Authority, true)
+                .WithTestLogging()
+                .WithCertificate(settings.GetCertificate())
+                .BuildConcrete();
+
+            Trace.WriteLine("Acquire a token from IDP.");
+            authResult = await confidentialApp
+                .AcquireTokenForClient(settings.AppScopes)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
+            Assert.IsNotNull(authResult.AccessToken);
+            Assert.IsTrue(authResult.AuthenticationResultMetadata.DurationTotalInMs > 50);
+
+            Trace.WriteLine("Acquire a token from cache.");
+            authResult = await confidentialApp
+                .AcquireTokenForClient(settings.AppScopes)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
+            Assert.IsNotNull(authResult.AccessToken);
+            Assert.AreEqual(TokenSource.Cache, authResult.AuthenticationResultMetadata.TokenSource);
+            Assert.AreEqual(CacheRefreshReason.NotApplicable, authResult.AuthenticationResultMetadata.CacheRefreshReason);
+            Assert.IsTrue(authResult.AuthenticationResultMetadata.DurationTotalInMs < 50);
```

Comment:
Added these to keep a check on latencies. Generally the cache read/write is about 0-20 ms for these tests. And the http ones are more than 100ms. There is enough margin. 
@bgavrilMS since we discussed having integration tests for the latencies, do you think these look good?

Reply:
I don't have a strong opinion here. 

AAD latency can be quite big when it has problems, so we can't add an assertion for when TokenSource is IdP.

For cached token, with this config, the latency should be way under 1ms.  So 50 isn't going to catch anything, except a major regression. 

How about we leave this in and if the test ever fails we remove it?

Reply:
Makes sense. Will remove for IDP and leave for the cache for now.